vignettes/porcelain.Rmd

```diff
@@ -0,0 +1,138 @@
+---
+title: "porcelain"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{porcelain}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+show_file <- function(path, lang) {
+  writeLines(c(paste0("```", lang), readLines(path), "```"))
+}
+porcelain_file <- function(path) {
+  system.file(path, package = "porcelain", mustWork = TRUE)
+}
+```
+
+`porcelain` is a package which allows creation of robust, testable HTTP API's from R. These can be used to expose any R calculation to other applications while communicating explicit expectations for inputs and outputs.  It is implemented on top of the excellent `plumber` package and uses `plumber`'s programmatic interface to build its API.  The interface, goals and tradeoffs are very different to plumber, but we think that this leads to high-quality APIs that are easy to put into production.
+
+## Motivation
+
+The `plumber` package makes it trivially easy to create APIs from R.  For example, from the [documentation](https://rplumber.io)
+
+```r
+#* Return the sum of two numbers
+#* @param a The first number to add
+#* @param b The second number to add
+#* @post /sum
+function(a, b) {
+  as.numeric(a) + as.numeric(b)
+}
+```
+
+Starting this API is easy,
+
+```r
+library(plumber)
+# 'plumber.R' is the location of the file shown above
+pr("plumber.R") %>%
+  pr_run(port=8000)
+```
+
+and we can interact with it using `curl`:
+
+```
+curl -w"\n" --data "a=4&b=3" "http://localhost:8000/sum"
+[7]
+```
+
+This simple example illustrates several issues that `porcelain` tries to solve:
+
+* The return value here is a json vector of length one even though the inputs were scalars - was that expected? How to we make this api more type-stable so that we can easily use it in upstream applications
```

Comment:
```suggestion
* The return value here is a json vector of length one even though the inputs were scalars - was that expected? How do we make this api more type-stable so that we can easily use it in upstream applications
```